internal/resources/azure/application_gateway.go

```diff
@@ -0,0 +1,192 @@
+package azure
+
+import (
+	"github.com/infracost/infracost/internal/resources"
+	"github.com/infracost/infracost/internal/schema"
+
+	"fmt"
+	"strings"
+
+	"github.com/infracost/infracost/internal/usage"
+	"github.com/shopspring/decimal"
+)
+
+type ApplicationGateway struct {
+	Address                string
+	Region                 string
+	SKUName                string
+	SKUCapacity            int64
+	MonthlyDataProcessedGB *float64 `infracost_usage:"monthly_data_processed_gb"`
+	MonthlyV2CapacityUnits *int64   `infracost_usage:"monthly_v2_capacity_units"`
+}
+
+var ApplicationGatewayUsageSchema = []*schema.UsageItem{{Key: "monthly_data_processed_gb", ValueType: schema.Float64, DefaultValue: 0}, {Key: "monthly_v2_capacity_units", ValueType: schema.Int64, DefaultValue: 0}}
+
+func (r *ApplicationGateway) PopulateUsage(u *schema.UsageData) {
+	resources.PopulateArgsWithUsage(r, u)
+}
+
+func (r *ApplicationGateway) BuildResource() *schema.Resource {
+	var monthlyDataProcessedGb, monthlyCapacityUnits *decimal.Decimal
+	var sku, tier string
+	costComponents := make([]*schema.CostComponent, 0)
+	tierLimits := []int{10240, 30720}
+
+	skuNameParts := strings.Split(r.SKUName, "_")
+	if len(skuNameParts[1]) != 0 {
+		sku = strings.ToLower(skuNameParts[1])
+	}
+	if sku != "v2" {
+		if strings.ToLower(skuNameParts[0]) == "standard" {
+			tier = "basic"
+		} else {
+			tier = "WAF"
+		}
+		costComponents = append(costComponents, r.gatewayCostComponent(fmt.Sprintf("Gateway usage (%s, %s)", tier, sku), tier, sku))
+
+		if r.MonthlyDataProcessedGB != nil {
+			monthlyDataProcessedGb = decimalPtr(decimal.NewFromFloat(*r.MonthlyDataProcessedGB))
+			result := usage.CalculateTierBuckets(*monthlyDataProcessedGb, tierLimits)
+
+			if sku == "small" {
+				if result[0].GreaterThan(decimal.Zero) {
+					costComponents = append(costComponents, r.dataProcessingCostComponent("Data processing (0-10TB)", sku, "0", &result[0]))
+				}
+				if result[1].GreaterThan(decimal.Zero) {
```

Comment:
Should we check the len here and the conditions below. I know it should be okay, but it feels better to be defensive here.

Reply:
Reverte